core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateUnnest.java

```diff
@@ -0,0 +1,402 @@
+ *     than 1 row for each input row. This is achieved by RowNumberNode partitioned by input rows (unique) + FilterNode.
+ * [3] If the original plan had LimitNode in the subquery, it has to be restored. It is achieved by RowNumberNode partitioned by input rows (unique)
+ *     and with maxRowCountPerPartition set to the limit.
+ * [4] All projections present in the subquery are restored on top of the rewritten UnnestNode. Apart from their original assignments,
+ *     they pass all the input symbols and the ordinality symbol which might be useful in the upstream plan (as in [1])
+ * [5] Type of the rewritten UnnestNode is LEFT with one exception: if both CorrelatedJoinNode and original UnnestNode are INNER, and there is no
+ *     EnforceSingleRowNode in the subquery, then the type is INNER. If the unnest type is rewritten from INNER to LEFT, the INNER semantics
+ *     is restored by a projection ([1]).
+ * </pre>
+ * Note: this rule captures and transforms different plans. Therefore there is some redundancy. Not every flow will use ordinality symbol
+ * or unique symbol. However, for simplicity, they are always present. It is up to other optimizer rules to prune unused symbols
+ * or redundant nodes.
+ */
+public class DecorrelateUnnest
+        implements Rule<CorrelatedJoinNode>
+{
+    private static final Pattern<CorrelatedJoinNode> PATTERN = correlatedJoin()
+            .with(nonEmpty(correlation()))
+            .with(filter().equalTo(TRUE_LITERAL))
+            .matching(node -> node.getType() == CorrelatedJoinNode.Type.INNER || node.getType() == CorrelatedJoinNode.Type.LEFT);
+
+    private final Metadata metadata;
+
+    public DecorrelateUnnest(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<CorrelatedJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Context context)
+    {
+        // determine shape of the subquery
+        PlanNode searchRoot = correlatedJoinNode.getSubquery();
+
+        // 1. find EnforceSingleRowNode in the subquery
+        Optional<EnforceSingleRowNode> enforceSingleRow = PlanNodeSearcher.searchFrom(searchRoot, context.getLookup())
+                .where(EnforceSingleRowNode.class::isInstance)
+                .recurseOnlyWhen(planNode -> false)
+                .findFirst();
+
+        if (enforceSingleRow.isPresent()) {
+            searchRoot = enforceSingleRow.get().getSource();
+        }
+
+        // 2. find LimitNode in the subquery
+        Optional<LimitNode> limitNode = PlanNodeSearcher.searchFrom(searchRoot, context.getLookup())
+                .where(LimitNode.class::isInstance)
+                .recurseOnlyWhen(ProjectNode.class::isInstance)
+                .findFirst();
+
+        if (limitNode.isPresent()) {
+            searchRoot = limitNode.get().getSource();
```

Comment:
do we care about value of `limitnode...getTiesResolvingScheme()` here?

Reply:
Good catch! I totally missed limit with ties. It can be supported as well as TopN.